src/utils/narrow.js

```diff
@@ -358,19 +359,29 @@ export const getNarrowsForMessage = (
  *
  * Chooses a topic narrow over a stream narrow.
  */
-export const getNarrowForReply = (message: Message | Outbox, ownEmail: string) => {
-  if (Array.isArray(message.display_recipient)) {
-    if (message.display_recipient.length > 1) {
-      return groupNarrow(
-        message.display_recipient.filter(x => x.email !== ownEmail).map(x => x.email),
-      );
-    } else if (message.display_recipient.length === 1) {
-      return privateNarrow(message.display_recipient[0].email);
-    } else {
-      // FIXME(crunchy): This shouldn't happen; reject at the edge.
-      return groupNarrow([]);
-    }
-  }
+export const getNarrowForReply = (message: Message | Outbox, ownEmail: string): Narrow => {
+  // TODO (?): We're saving callers the trouble of giving us the
+  // message's flags (which may be missing on `message`), since there
+  // aren't any flags that could influence where a message's reply
+  // should go.
+  const fakeFlags = [];
+
+  // We can't send to, e.g., the "starred" narrow.
+  const sendableNarrows = getNarrowsForMessage(message, ownEmail, fakeFlags).filter(
+    canSendToNarrow,
+  );
 
-  return topicNarrow(message.display_recipient, message.subject);
+  const [sendableStreamNarrow, sendableTopicNarrow] = [
+    sendableNarrows.find(n => isStreamNarrow(n)),
+    sendableNarrows.find(n => isTopicNarrow(n)),
+  ];
+  if (sendableStreamNarrow && sendableTopicNarrow) {
+    return sendableTopicNarrow;
+  } else if (sendableNarrows[0]) {
+    return sendableNarrows[0];
```

Comment:
This feels pretty fiddly to me. I think a cleaner way to avoid the duplication here is to factor just the duplicated code out: effectively this version is calling `getNarrowsForMessage` but then undoing some of what that function did, and we can get a cleaner result by instead having both functions call a common simpler helper that just does the part that's duplicated.

Following #4239, `pmKeyRecipientsFromMessage` is that helper and that's what  `getNarrowFromMessage` now uses.

Reply:
> Following #4239, `pmKeyRecipientsFromMessage` is that helper and that's what `getNarrowFromMessage` now uses.

(#4329, I think. 🙂)

Reply:
> I think a cleaner way to avoid the duplication here is to factor just the duplicated code out: effectively this version is calling `getNarrowsForMessage` but then undoing some of what that function did, and we can get a cleaner result by instead having both functions call a common simpler helper that just does the part that's duplicated.

Ooh, nice! Thanks.